language/move-lang/src/cfgir/inline_blocks.rs

```diff
@@ -85,5 +89,37 @@ fn inline_single_target_blocks(
             }
         }
     }
+
+    let changed = !remapping.is_empty();
+    remap_to_last_target(remapping, start, finished_blocks);
     changed
 }
+
+/// In order to preserve loop invariants at the bytecode level, when a block is "inlined", that
+/// block needs to be relabelled as the "inlined" block
+/// For example:
+/// Before:
+///   A: block_a; jump B
+///   B: block_b
+///
+///   s.t. A is the only one jumping to B
+///
+/// After:
+///   B: block_a; block_b
+fn remap_to_last_target(
+    mut remapping: BTreeMap<Label, Label>,
+    start: Label,
+    blocks: &mut BasicBlocks,
+) {
+    remapping.remove(&start);
```

Comment:
It might be worth a comment here to mention why the start block doesn't need to be relabeled.

Reply:
It's more that it *can't* be relabeled due to limitations on the current cfg api. I'll add a comment